src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java

```diff
@@ -769,25 +772,49 @@ private static GcvScope getScopeRecursively(
             return cached.get();
         }
 
-        GcvScope gcvScope = component.getDependents().stream()
-                .filter(dep -> !dep.isConstraint())
-                .filter(dep -> dep.getRequested() instanceof ModuleComponentSelector)
-                .map(dependent -> {
-                    ModuleIdentifier requestedModule =
-                            ((ModuleComponentSelector) dependent.getRequested()).getModuleIdentifier();
-                    // If the dependency came from a project, then the requested ModuleIdentifier should be in the
-                    // edgeScopes. Otherwise, recurse until we find a project dependent.
-                    Optional<GcvScope> maybeScope = directDependencyScopes.getScopeFor(requestedModule);
-                    if (dependent.getFrom().getId() instanceof ProjectComponentIdentifier && maybeScope.isPresent()) {
-                        return maybeScope.get();
-                    }
-                    return getScopeRecursively(dependent.getFrom(), scopeCache, directDependencyScopes);
-                })
+        Set<ResolvedDependencyResult> traversedComponents = new HashSet<>();
+        Deque<ResolvedDependencyResult> stack =
+                new ArrayDeque<>(component.getDependents().size());
+        stack.addAll(component.getDependents());
+
+        Set<GcvScope> discoveredScopes = new HashSet<>(2);
+        while (!stack.isEmpty()) {
+            ResolvedDependencyResult dependent = stack.removeFirst();
+            if (dependent.isConstraint()
+                    || !(dependent.getRequested() instanceof ModuleComponentSelector)
+                    || traversedComponents.contains(dependent)) {
+                continue;
+            }
+
+            Optional<GcvScope> cachedValue = Optional.ofNullable(scopeCache.get(dependent.getFrom()));
```

Comment:
`dependent.getFrom()` is not the same as `((ModuleComponentSelector) dependent.getRequested()).getModuleIdentifier()`

for a dependency `org:foo:1.0 -> org:bar:1.1`
* `.from` returns the `ResolvedComponentResult` `org:foo:1.0`
* `.requested.moduleIdentifier` returns the `ModuleIdentifier` `org:bar`

so you're putting the wrong type into scopeCache.get

Reply:
orr.. never mind, I got myself confused with these types and what was happening before